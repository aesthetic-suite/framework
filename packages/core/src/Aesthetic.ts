/* eslint-disable no-param-reassign, lines-between-class-members, no-dupe-class-members */

import { FontFace as SSSFontFace, formatFontFace, LocalBlock } from '@aesthetic/sss';
import { Theme, ThemeRegistry } from '@aesthetic/system';
import {
  ClassName,
  Direction,
  Engine,
  FontFace,
  Keyframes,
  RenderOptions,
  ThemeName,
} from '@aesthetic/types';
import { arrayLoop, createState, isDOM } from '@aesthetic/utils';
import StyleSheet from './StyleSheet';
import {
  AestheticOptions,
  EventListener,
  EventType,
  GlobalSheet,
  GlobalSheetFactory,
  LocalSheet,
  LocalSheetFactory,
  OnChangeDirection,
  OnChangeTheme,
  RenderResultSheet,
  SheetParams,
} from './types';

export default class Aesthetic<Result = ClassName, Block extends object = LocalBlock> {
  protected activeDirection = createState<Direction>();

  protected activeTheme = createState<ThemeName>();

  protected globalSheetRegistry = new Map<ThemeName, GlobalSheet<unknown, Block, Result>>();

  protected listeners = new Map<EventType, Set<EventListener>>();

  protected options: AestheticOptions = {};

  protected styleEngine = createState<Engine<Result>>();

  protected themeRegistry = new ThemeRegistry<Block>();

  constructor(options?: AestheticOptions) {
    if (options) {
      this.configure(options);
    }

    // Must be bound manually because of method overloading
    this.emit = this.emit.bind(this);
    this.subscribe = this.subscribe.bind(this);
    this.unsubscribe = this.unsubscribe.bind(this);
  }

  /**
   * Change the active direction.
   */
  changeDirection = (direction: Direction, propagate: boolean = true): void => {
    if (direction === this.activeDirection.get()) {
      return;
    }

    const engine = this.getEngine();

    // Set the active direction
    this.activeDirection.set(direction);

    // Update engine direction
    engine.direction = direction;
    engine.setDirection(direction);

    // Let consumers know about the change
    if (propagate) {
      this.emit('change:direction', [direction]);
    }
  };

  /**
   * Change the active theme.
   */
  changeTheme = (name: ThemeName, propagate: boolean = true): void => {
    if (name === this.activeTheme.get()) {
      return;
    }

    const theme = this.getTheme(name);
    const engine = this.getEngine();

    // Set the active theme
    this.activeTheme.set(name);

    // Set root variables
    if (this.options.rootVariables) {
      engine.setRootVariables(theme.toVariables());
    }

    // Render theme styles
    const results = this.renderThemeStyles(theme);

    // Update engine theme
    engine.setTheme(results);

    // Let consumers know about the change
    if (propagate) {
      this.emit('change:theme', [name, results]);
    }
  };

  /**
   * Configure options unique to this instance.
   */
  configure = (customOptions: AestheticOptions): void => {
    Object.assign(this.options, customOptions);

    // Configure the engine with itself to reapply options
    const engine = this.styleEngine.get();

    if (engine) {
      this.configureEngine(engine);
    }
  };

  /**
   * Configuring which styling engine to use.
   */
  configureEngine = (engine: Engine<Result>): void => {
    const { options } = this;

    if (!engine.directionConverter && options.directionConverter) {
      engine.directionConverter = options.directionConverter;
    }

    if (!engine.unitSuffixer && options.defaultUnit) {
      engine.unitSuffixer = options.defaultUnit;
    }

    if (!engine.vendorPrefixer && options.vendorPrefixer) {
      engine.vendorPrefixer = options.vendorPrefixer;
    }

    this.styleEngine.set(engine);
  };

  /**
   * Create a local style sheet for use within components.
   */
  createComponentStyles = <T = unknown>(
    factory: LocalSheetFactory<T, Block>,
  ): LocalSheet<T, Block, Result> => new StyleSheet('local', factory);

  /**
   * Create a global style sheet for root theme styles.
   */
  createThemeStyles = <T = unknown>(
    factory: GlobalSheetFactory<T, Block>,
  ): GlobalSheet<T, Block, Result> => new StyleSheet('global', factory);

  /**
   * Emit all listeners by type, with the defined arguments.
   */
  emit(type: 'change:direction', args: Parameters<OnChangeDirection>): void;
  emit(type: 'change:theme', args: Parameters<OnChangeTheme>): void;
  emit(type: EventType, args: unknown[]): void {
    this.getListeners(type).forEach((listener) => {
      listener(...args);
    });
  }

  /**
   * Generate a class name using the selectors of a style sheet.
   * If an object is provided, it will be used to check for variants.
   */
  generateClassName = <T extends string>(
    keys: T[],
    variants: string[],
    classNames: RenderResultSheet<ClassName>,
  ): ClassName => {
    let className = '';

    arrayLoop(keys, (key) => {
      const hash = classNames[key];

      if (!hash) {
        return;
      }

      if (hash.result) {
        className += ' ';
        className += hash.result;
      }

      if (hash.variants) {
        arrayLoop(variants, (variant) => {
          if (hash.variants?.[variant]) {
            className += ' ';
            className += hash.variants[variant];
          }
        });
      }
    });

    return className.trim();
  };

  /**
   * Return the active direction for the entire application. If an active direction is undefined,
   * it will be detected from the browser's `dir` attribute.
   */
  getActiveDirection = (): Direction => {
    const active = this.activeDirection.get();

    if (active) {
      return active;
    }

    // Detect theme from engine
    const direction: Direction = this.getEngine().direction || 'ltr';

    this.changeDirection(direction);

    return direction;
  };

  /**
   * Return the currently active theme instance. If an active instance has not been defined,
   * one will be detected from the client's browser preferences.
   */
  getActiveTheme = (): Theme<Block> => {
    const active = this.activeTheme.get();

    if (active) {
      return this.getTheme(active);
    }

    // Detect theme from browser preferences
    const engine = this.getEngine();
    const theme = this.themeRegistry.getPreferredTheme({
      matchColorScheme: engine.prefersColorScheme,
      matchContrastLevel: engine.prefersContrastLevel,
    });

    this.changeTheme(theme.name);

    return theme;
  };

  /**
   * Return the current style engine.
   */
  getEngine = (): Engine<Result> => {
    const current = global.AESTHETIC_CUSTOM_ENGINE || this.styleEngine.get();

    if (current) {
      return current;
    }

    throw new Error('No style engine defined. Have you configured one with `configureEngine()`?');
  };

  /**
   * Return a set of listeners, or create it if it does not exist.
   */
  getListeners = <T extends Function>(type: EventType): Set<T> => {
    const set = this.listeners.get(type) || new Set();

    this.listeners.set(type, set);

    return (set as unknown) as Set<T>;
  };

  /**
   * Return a theme instance by name.
   */
  getTheme = (name: ThemeName): Theme<Block> => this.themeRegistry.getTheme(name);

  /**
   * Register a theme, with optional global theme styles.
   */
  registerTheme = (
    name: ThemeName,
    theme: Theme<Block>,
    sheet: GlobalSheet<unknown, Block, Result> | null = null,
    isDefault: boolean = false,
  ): void => {
    this.themeRegistry.register(name, theme, isDefault);

    if (sheet) {
      if (__DEV__) {
        if (!(sheet instanceof StyleSheet) || sheet.type !== 'global') {
          throw new TypeError('Rendering theme styles require a `GlobalSheet` instance.');
        }
      }

      this.globalSheetRegistry.set(name, sheet);
    }
  };

  /**
   * Register a default light or dark theme, with optional global theme styles.
   */
  registerDefaultTheme = (
    name: ThemeName,
    theme: Theme<Block>,
    sheet: GlobalSheet<unknown, Block, Result> | null = null,
  ): void => {
    this.registerTheme(name, theme, sheet, true);
  };

  /**
   * Render a component style sheet to the document with the defined style query parameters.
   */
  renderComponentStyles = <T = unknown>(
    sheet: LocalSheet<T, Block, Result>,
    themeName?: ThemeName,
  ) => {
    if (__DEV__) {
      if (!(sheet instanceof StyleSheet) || sheet.type !== 'local') {
        throw new TypeError('Rendering component styles require a `LocalSheet` instance.');
      }
    }

    const theme = themeName ? this.getTheme(themeName) : this.getActiveTheme();

    return sheet.render(this.getEngine(), theme, {
      customProperties: this.options.customProperties,
      deterministic: this.options.deterministicClasses,
      direction: this.getActiveDirection(),
      vendor: !!this.options.vendorPrefixer,
    });
  };

  /**
   * Render a `@font-face` to the global style sheet and return the font family name.
   */
  renderFontFace = (
    fontFace: FontFace | SSSFontFace,
    fontFamily?: string,
    params?: RenderOptions,
  ): string =>
    this.getEngine().renderFontFace(
      formatFontFace({
        fontFamily,
        ...fontFace,
      }) as FontFace,
      params,
    );

  /**
   * Render an `@import` to the global style sheet and return the import path.
   */
  renderImport = (path: string, params?: RenderOptions): string =>
    this.getEngine().renderImport(path, params);

  /**
   * Render a `@keyframes` to the global style sheet and return the animation name.
   */
  renderKeyframes = (
    keyframes: Keyframes,
    animationName?: string,
    params?: RenderOptions,
  ): string => this.getEngine().renderKeyframes(keyframes, animationName, params);

  /**
   * Render a global theme style sheet and return a result, if one was generated.
   */
  renderThemeStyles = (theme: Theme<Block>, params: SheetParams = {}): Result[] => {
    const sheet = this.globalSheetRegistry.get(theme.name);
    const results: Result[] = [];

    // Render theme CSS variables
    if (isDOM()) {
      results.push(this.getEngine().renderRuleGrouped(theme.toVariables(), { type: 'global' }));
    }

    // Render theme styles
    if (sheet) {
      const result = sheet.render(this.getEngine(), theme, {
        customProperties: this.options.customProperties,
        direction: this.getActiveDirection(),
        vendor: !!this.options.vendorPrefixer,
        ...params,
      })['@root']?.result;

      if (result) {
        results.push(result);
      }
    }

    return results;
  };

  /**
   * Subscribe and listen to an event by name.
   */
  subscribe(type: 'change:direction', listener: OnChangeDirection): () => void;
  subscribe(type: 'change:theme', listener: OnChangeTheme): () => void;
  subscribe(type: EventType, listener: Function): () => void {
    this.getListeners(type).add(listener);

    return () => {
      this.unsubscribe(type as 'change:theme', listener as OnChangeTheme);
    };
  }

  /**
   * Unsubscribe from an event by name.
   */
  unsubscribe(type: 'change:direction', listener: OnChangeDirection): void;
  unsubscribe(type: 'change:theme', listener: OnChangeTheme): void;
  unsubscribe(type: EventType, listener: Function): void {
    this.getListeners(type).delete(listener);
  }
}
