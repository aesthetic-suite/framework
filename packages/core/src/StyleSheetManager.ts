export default class StyleSheetManager {
  private element: HTMLStyleElement;

  private sheet: CSSStyleSheet;

  constructor() {
    this.element = document.createElement('style');
    this.element.type = 'text/css';
    this.element.media = 'screen';
    this.element.setAttribute('data-aesthetic', 'true');

    document.head.append(this.element);

    // This must happen after the element is inserted into the DOM,
    // otherwise the value is null.
    this.sheet = this.element.sheet as CSSStyleSheet;
  }

  getInjectedStyles(): string {
    return (
      (this.element.textContent || '') +
      Array.from(this.sheet.cssRules)
        .map(rule => rule.cssText)
        .join('')
    );
  }

  injectRule(rule: string): this {
    this.sheet.insertRule(rule, this.sheet.cssRules.length);

    return this;
  }

  injectStatements(css: string): this {
    this.element.textContent += css;

    return this;
  }
}
