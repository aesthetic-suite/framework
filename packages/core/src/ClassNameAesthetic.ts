import Aesthetic from './Aesthetic';
import { ClassName } from './types';

export default class ClassNameAesthetic<Theme extends object> extends Aesthetic<
  Theme,
  any,
  ClassName
> {
  protected transformToClassName(styles: any[]): ClassName {
    return styles.filter(style => style && typeof style === 'string').join(' ');
  }
}
