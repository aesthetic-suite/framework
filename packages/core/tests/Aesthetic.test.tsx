import { StyleSheetTestUtils } from 'aphrodite';
import { createRenderer } from 'fela';
import webPreset from 'fela-preset-web';
import { create } from 'jss';
// @ts-ignore
import preset from 'jss-preset-default';
import { TypeStyle } from 'typestyle';
import AphroditeAesthetic from 'aesthetic-adapter-aphrodite';
import CSSModulesAesthetic from 'aesthetic-adapter-css-modules';
import FelaAesthetic from 'aesthetic-adapter-fela';
import JSSAesthetic from 'aesthetic-adapter-jss';
import TypeStyleAesthetic from 'aesthetic-adapter-typestyle';
import Aesthetic from '../src/Aesthetic';
import ClassNameAesthetic from '../src/ClassNameAesthetic';
import StyleSheetManager from '../src/StyleSheetManager';
import {
  TestAesthetic,
  registerTestTheme,
  SYNTAX_GLOBAL,
  SYNTAX_UNIFIED_LOCAL_FULL,
} from '../src/testUtils';

describe('Aesthetic', () => {
  let instance: Aesthetic<any, any, any>;

  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();

    instance = new TestAesthetic();

    registerTestTheme(instance);
  });

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  describe('constructor()', () => {
    it('merges options', () => {
      instance = new TestAesthetic({
        stylesPropName: 'styleSheet',
      });

      expect(instance.options).toEqual({
        cxPropName: 'cx',
        extendable: false,
        passThemeProp: false,
        pure: true,
        rtl: false,
        stylesPropName: 'styleSheet',
        theme: 'default',
        themePropName: 'theme',
      });
    });
  });

  describe('applyGlobalStyles()', () => {
    it('does nothing if already applied', () => {
      const spy = jest.spyOn(instance, 'flushStyles');

      // @ts-ignore Allow override
      instance.appliedGlobals = true;
      instance.applyGlobalStyles({});

      expect(spy).not.toHaveBeenCalled();
    });

    it('does nothing if no global styles defined for theme', () => {
      // @ts-ignore Allow access
      const spy = jest.spyOn(instance, 'processStyleSheet');

      instance.globals.default = null;
      instance.applyGlobalStyles({});

      expect(spy).not.toHaveBeenCalled();
    });

    it('processes global styles if defined for a theme', () => {
      // @ts-ignore Allow access
      const spy = jest.spyOn(instance, 'processStyleSheet');

      instance.applyGlobalStyles({});

      expect(spy).toHaveBeenCalledWith({}, ':root');
    });
  });

  // Will have no properties as no unified syntax handlers are defined
  describe('createStyleSheet()', () => {
    beforeEach(() => {
      // @ts-ignore Allow access
      instance.setStyleSheet('foo', ({ unit }) => ({
        el: {
          display: 'block',
          padding: unit,
          color: 'black',
        },
      }));
    });

    it('returns the style sheet', () => {
      expect(instance.createStyleSheet('foo', {})).toEqual({
        el: {},
      });
    });

    it('calls `convertStyleSheet` for unified syntax while passing theme', () => {
      const spy = jest.spyOn(instance.syntax, 'convertStyleSheet');

      instance.createStyleSheet('foo', {});

      expect(spy).toHaveBeenCalledWith(
        {
          el: {
            display: 'block',
            padding: 8,
            color: 'black',
          },
        },
        { name: 'foo' },
      );
    });

    it('calls `processStyleSheet` with converted syntax', () => {
      // @ts-ignore Allow access
      const spy = jest.spyOn(instance, 'processStyleSheet');

      instance.createStyleSheet('foo', {});

      expect(spy).toHaveBeenCalledWith({ el: {} }, 'foo');
    });

    it('calls `applyGlobalStyles`', () => {
      // @ts-ignore Allow access
      const spy = jest.spyOn(instance, 'applyGlobalStyles');

      instance.createStyleSheet('foo', {});

      expect(spy).toHaveBeenCalled();
    });

    it('caches the result', () => {
      expect(instance.cache.foo).toBeUndefined();

      instance.createStyleSheet('foo', {});

      expect(instance.cache.foo).toEqual({ el: {} });

      const sheet = instance.createStyleSheet('foo', {});

      expect(instance.cache.foo).toEqual(sheet);
    });
  });

  describe('extendStyles()', () => {
    it('returns a stylesheet', () => {
      const styleSheet = instance.extendStyles(() => ({ element: { display: 'block' } }));

      expect(typeof styleSheet).toBe('function');
      expect(styleSheet({})).toEqual({ element: { display: 'block' } });
    });

    it('passes theme variables', () => {
      const styleSheet = instance.extendStyles(theme => ({ element: { padding: theme.unit } }));

      expect(styleSheet({ unit: 8 })).toEqual({ element: { padding: 8 } });
    });

    it('deep merges multiple stylesheets', () => {
      const styleSheet = instance.extendStyles(
        theme => ({ element: { padding: theme.unit } }),
        () => ({ element: { display: 'block' } }),
        ({ color }) => ({ elementActive: { color } }),
      );

      expect(styleSheet({ unit: 8, color: 'red' })).toEqual({
        element: { padding: 8, display: 'block' },
        elementActive: { color: 'red' },
      });
    });
  });

  describe('extendTheme()', () => {
    it('errors if the parent theme doesnt exist', () => {
      expect(() => instance.extendTheme('bar', 'foo', {})).toThrowErrorMatchingSnapshot();
    });

    it('deep merges the parent and child theme', () => {
      instance.registerTheme('foo', {
        unit: 'px',
        unitSize: 8,
        colors: {
          primary: 'red',
        },
      });

      instance.extendTheme('bar', 'foo', {
        unit: 'em',
        colors: {
          primary: 'blue',
          secondary: 'orange',
        },
      });

      expect(instance.themes.bar).toEqual({
        unit: 'em',
        unitSize: 8,
        colors: {
          primary: 'blue',
          secondary: 'orange',
        },
      });
    });

    it('inherits parent theme global styles', () => {
      const spy = jest.fn();

      instance.registerTheme('foo', {}, spy);
      instance.extendTheme('bar', 'foo', {});

      expect(instance.globals.bar).toBe(spy);
    });
  });

  describe('getStyleSheet()', () => {
    beforeEach(() => {
      // @ts-ignore Allow access
      instance.setStyleSheet('foo', () => ({
        el: { display: 'block' },
      }));
    });

    it('errors if no theme', () => {
      instance.options.theme = 'unknown';

      expect(() => {
        // @ts-ignore Allow access
        instance.getStyleSheet('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('returns the style sheet', () => {
      // @ts-ignore Allow access
      expect(instance.getStyleSheet('foo')).toEqual({
        el: { display: 'block' },
      });
    });

    it('passes theme to the style callback', () => {
      // @ts-ignore Allow access
      instance.setStyleSheet('bar', theme => ({
        el: { padding: theme.unit * 2 },
      }));

      // @ts-ignore Allow access
      expect(instance.getStyleSheet('bar')).toEqual({
        el: { padding: 16 },
      });
    });

    it('inherits styles from parent', () => {
      // @ts-ignore Allow access
      instance.setStyleSheet('bar', () => ({
        el: {
          color: 'red',
          ':hover': {
            color: 'darkred',
          },
        },
      }));

      // @ts-ignore Allow access
      instance.setStyleSheet(
        'baz',
        () => ({
          el: {
            background: 'blue',
            ':hover': {
              color: 'green',
            },
          },
        }),
        'bar',
      );

      // @ts-ignore Allow access
      instance.setStyleSheet(
        'qux',
        () => ({
          el: { display: 'block' },
        }),
        'baz',
      );

      // @ts-ignore Allow access
      expect(instance.getStyleSheet('bar')).toEqual({
        el: {
          color: 'red',
          ':hover': {
            color: 'darkred',
          },
        },
      });

      // @ts-ignore Allow access
      expect(instance.getStyleSheet('baz')).toEqual({
        el: {
          color: 'red',
          background: 'blue',
          ':hover': {
            color: 'green',
          },
        },
      });

      // @ts-ignore Allow access
      expect(instance.getStyleSheet('qux')).toEqual({
        el: {
          color: 'red',
          background: 'blue',
          display: 'block',
          ':hover': {
            color: 'green',
          },
        },
      });
    });
  });

  describe('getStyleSheetManager', () => {
    it('returns and sets a style sheet manager', () => {
      // @ts-ignore Allow access
      expect(instance.sheetManager).toBeNull();

      // @ts-ignore Allow access
      const manager = instance.getStyleSheetManager();

      // @ts-ignore Allow access
      expect(instance.sheetManager).not.toBeNull();
      expect(manager).toBeInstanceOf(StyleSheetManager);
    });
  });

  describe('getTheme()', () => {
    it('errors if the theme doesnt exist', () => {
      expect(() => {
        // @ts-ignore Allow access
        instance.getTheme('foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if the theme is not an object', () => {
      instance.themes.default = 123;

      expect(() => {
        // @ts-ignore Allow access
        instance.getTheme();
      }).toThrowErrorMatchingSnapshot();
    });

    it('returns the default theme if no name provided', () => {
      // @ts-ignore Allow access
      expect(instance.getTheme()).toEqual({ unit: 8 });
    });

    it('returns the theme by name', () => {
      // @ts-ignore Allow access
      expect(instance.getTheme('default')).toEqual({ unit: 8 });
    });
  });

  describe('processStyleSheet()', () => {
    it('returns the style sheet as an object', () => {
      const sheet = { el: {} };
      // @ts-ignore Allow access
      const styleSheet = instance.processStyleSheet(sheet, 'styleName');

      expect(sheet).toEqual(styleSheet);
    });
  });

  describe('registerTheme()', () => {
    it('errors if a theme name has been used', () => {
      expect(() => instance.registerTheme('default', {})).toThrowErrorMatchingSnapshot();
    });

    it('errors if a theme style is not an object', () => {
      expect(() => instance.registerTheme('foo', 123)).toThrowErrorMatchingSnapshot();
    });

    it('errors if global styles is not an object', () => {
      // @ts-ignore Allow non-object
      expect(() => instance.registerTheme('foo', {}, 123)).toThrowErrorMatchingSnapshot();
    });

    it('registers theme and sets global styles', () => {
      expect(instance.globals.foo).toBeUndefined();

      instance.registerTheme('foo', { unitSize: 6 }, () => SYNTAX_GLOBAL);

      expect(instance.themes.foo).toEqual({ unitSize: 6 });

      expect(instance.globals.foo).toBeDefined();
    });
  });

  describe('setStyleSheet()', () => {
    it('errors if styles have been set', () => {
      instance.styles.foo = () => ({});

      expect(() => {
        // @ts-ignore Allow access
        instance.setStyleSheet('foo', () => ({}));
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if styles try to extend from itself', () => {
      instance.styles.foo = () => ({});

      expect(() => {
        // @ts-ignore Allow access
        instance.setStyleSheet('foo', () => ({}), 'foo');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if styles are not a function', () => {
      // @ts-ignore Allow invalid type
      expect(() => instance.setStyleSheet('foo', 123)).toThrowErrorMatchingSnapshot();
      // @ts-ignore Allow invalid type
      expect(() => instance.setStyleSheet('foo', 'abc')).toThrowErrorMatchingSnapshot();
      // @ts-ignore Allow invalid type
      expect(() => instance.setStyleSheet('foo', [])).toThrowErrorMatchingSnapshot();
      // @ts-ignore Allow invalid type
      expect(() => instance.setStyleSheet('foo', true)).toThrowErrorMatchingSnapshot();
    });

    it('errors if extended styles do not exist', () => {
      expect(() =>
        // @ts-ignore Allow access
        instance.setStyleSheet('foo', () => ({}), 'parent'),
      ).toThrowErrorMatchingSnapshot();
    });

    it('sets styles', () => {
      expect(instance.styles.foo).toBeUndefined();

      // @ts-ignore Allow access
      instance.setStyleSheet('foo', () => ({
        header: { color: 'red' },
        footer: { padding: 5 },
      }));

      expect(instance.styles.foo).toBeDefined();
    });

    it('sets styles and extends parents', () => {
      expect(instance.styles.bar).toBeUndefined();
      expect(instance.parents.bar).toBeUndefined();

      // @ts-ignore Allow access
      instance.setStyleSheet('foo', () => ({
        header: { color: 'red' },
        footer: { padding: 5 },
      }));

      // @ts-ignore Allow access
      instance.setStyleSheet(
        'bar',
        () => ({
          child: { margin: 5 },
        }),
        'foo',
      );

      expect(instance.parents.bar).toBe('foo');
    });
  });

  describe('transformStyles()', () => {
    it('errors for invalid value', () => {
      expect(() => {
        instance.transformStyles(
          // @ts-ignore Allow invalid type
          [123],
          {},
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('combines strings into a class name', () => {
      expect(instance.transformStyles(['foo', 'bar'], {})).toBe('foo bar');
    });

    it('calls `transformToClassName` method', () => {
      // @ts-ignore Allow access
      const spy = jest.spyOn(instance, 'transformToClassName');

      instance.transformStyles([{ color: 'red' }, { display: 'block' }], {});

      expect(spy).toHaveBeenCalledWith([{ color: 'red' }, { display: 'block' }]);
    });

    it('ignores falsey values', () => {
      expect(instance.transformStyles([null, false, 0, '', undefined], {})).toBe('');
    });

    it('strips period prefix', () => {
      expect(instance.transformStyles(['.foo', 'bar .qux'], {})).toBe('foo bar qux');
    });

    it('handles expression values', () => {
      expect(instance.transformStyles(['foo', true && 'bar', 5 > 10 && 'baz'], {})).toBe('foo bar');
    });

    it('joins strings', () => {
      expect(instance.transformStyles(['foo', '123', 'bar'], {})).toBe('foo 123 bar');
    });
  });

  describe('adapters', () => {
    it('supports class names', () => {
      const adapter = new ClassNameAesthetic();
      adapter.registerTheme('default', {});
      // @ts-ignore Allow access
      adapter.setStyleSheet('foo', () => ({ button: 'button' }));

      const styleSheet = adapter.createStyleSheet('foo', {});

      expect(adapter.transformStyles([styleSheet.button], {})).toMatchSnapshot();
    });

    it('supports Aphrodite', () => {
      const adapter = new AphroditeAesthetic();
      adapter.registerTheme('default', {});
      // @ts-ignore Allow access
      adapter.setStyleSheet('foo', () => SYNTAX_UNIFIED_LOCAL_FULL as any);

      const styleSheet = adapter.createStyleSheet('foo', {});

      expect(adapter.transformStyles([styleSheet.button], {})).toMatchSnapshot();
    });

    it('supports CSS modules', () => {
      const adapter = new CSSModulesAesthetic();
      adapter.registerTheme('default', {});
      // @ts-ignore Allow access
      adapter.setStyleSheet('foo', () => ({ button: 'button' }));

      const styleSheet = adapter.createStyleSheet('foo', {});

      expect(adapter.transformStyles([styleSheet.button], {})).toMatchSnapshot();
    });

    it('supports Fela', () => {
      const adapter = new FelaAesthetic(
        createRenderer({
          plugins: [...webPreset],
        }),
      );
      adapter.registerTheme('default', {});
      // @ts-ignore Allow access
      adapter.setStyleSheet('foo', () => SYNTAX_UNIFIED_LOCAL_FULL as any);

      const styleSheet = adapter.createStyleSheet('foo', {});

      expect(adapter.transformStyles([styleSheet.button], {})).toMatchSnapshot();
    });

    it('supports JSS', () => {
      const jss = create();
      jss.setup(preset());

      const adapter = new JSSAesthetic(jss);
      adapter.registerTheme('default', {});
      // @ts-ignore Allow access
      adapter.setStyleSheet('foo', () => SYNTAX_UNIFIED_LOCAL_FULL as any);

      const styleSheet = adapter.createStyleSheet('foo', {});

      expect(adapter.transformStyles([styleSheet.button], {})).toMatchSnapshot();
    });

    it('supports TypeStyle', () => {
      const adapter = new TypeStyleAesthetic(new TypeStyle({ autoGenerateTag: false }));
      adapter.registerTheme('default', {});
      // @ts-ignore Allow access
      adapter.setStyleSheet('foo', () => SYNTAX_UNIFIED_LOCAL_FULL as any);

      const styleSheet = adapter.createStyleSheet('foo', {});

      expect(adapter.transformStyles([styleSheet.button], {})).toMatchSnapshot();
    });
  });
});
