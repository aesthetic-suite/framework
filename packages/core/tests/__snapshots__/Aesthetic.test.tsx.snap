// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aesthetic adapters supports Aphrodite 1`] = `"button_17ap0k4"`;

exports[`Aesthetic adapters supports CSS modules 1`] = `"button"`;

exports[`Aesthetic adapters supports Fela 1`] = `"a b c d e f g h i j k l m n o p q r s t u v w x"`;

exports[`Aesthetic adapters supports JSS 1`] = `"button-0-1-1"`;

exports[`Aesthetic adapters supports TypeStyle 1`] = `"f1lp6j3"`;

exports[`Aesthetic adapters supports class names 1`] = `"button"`;

exports[`Aesthetic extendTheme() errors if the parent theme doesnt exist 1`] = `"Theme \\"foo\\" does not exist."`;

exports[`Aesthetic getStyleSheet() errors if no theme 1`] = `"Theme \\"unknown\\" does not exist."`;

exports[`Aesthetic getTheme() errors if the theme doesnt exist 1`] = `"Theme \\"foo\\" does not exist."`;

exports[`Aesthetic getTheme() errors if the theme is not an object 1`] = `"Theme \\"default\\" does not exist."`;

exports[`Aesthetic registerTheme() errors if a theme name has been used 1`] = `"Theme \\"default\\" already exists."`;

exports[`Aesthetic registerTheme() errors if a theme style is not an object 1`] = `"Theme \\"foo\\" must be a style object."`;

exports[`Aesthetic registerTheme() errors if global styles is not an object 1`] = `"Definition for \\"foo\\" must be null or a function."`;

exports[`Aesthetic setStyleSheet() errors if extended styles do not exist 1`] = `"Cannot extend from \\"parent\\" as those styles do not exist."`;

exports[`Aesthetic setStyleSheet() errors if styles are not a function 1`] = `"Definition for \\"foo\\" must be null or a function."`;

exports[`Aesthetic setStyleSheet() errors if styles are not a function 2`] = `"Definition for \\"foo\\" must be null or a function."`;

exports[`Aesthetic setStyleSheet() errors if styles are not a function 3`] = `"Definition for \\"foo\\" must be null or a function."`;

exports[`Aesthetic setStyleSheet() errors if styles are not a function 4`] = `"Definition for \\"foo\\" must be null or a function."`;

exports[`Aesthetic setStyleSheet() errors if styles have been set 1`] = `"Styles have already been defined for \\"foo\\"."`;

exports[`Aesthetic setStyleSheet() errors if styles try to extend from itself 1`] = `"Cannot extend styles from itself."`;

exports[`Aesthetic transformStyles() errors for invalid value 1`] = `"Unsupported style type to transform."`;
